src/sinks/prometheus/sink.rs

```diff
@@ -0,0 +1,291 @@
+use crate::{
+    buffers::Acker,
+    config::{DataType, GenerateConfig, Resource, SinkConfig, SinkContext, SinkDescription},
+    event::metric::MetricKind,
+    sinks::{
+        util::{statistic::validate_quantiles, MetricEntry, StreamSink},
+        Healthcheck, VectorSink,
+    },
+    Event,
+};
+use async_trait::async_trait;
+use chrono::Utc;
+use futures::{future, stream::BoxStream, FutureExt, StreamExt, TryFutureExt};
+use hyper::{
+    header::HeaderValue,
+    service::{make_service_fn, service_fn},
+    Body, Method, Request, Response, Server, StatusCode,
+};
+use indexmap::IndexSet;
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+use std::{
+    collections::HashSet,
+    convert::Infallible,
+    net::SocketAddr,
+    sync::{Arc, RwLock},
+};
+use stream_cancel::{Trigger, Tripwire};
+
+use super::collector::{self, MetricCollector as _};
+
+const MIN_FLUSH_PERIOD_SECS: u64 = 1;
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("Flush period for sets must be greater or equal to {} secs", min))]
+    FlushPeriodTooShort { min: u64 },
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct PrometheusSinkConfig {
```

Comment:
It's a little confusing to have a prometheus sink and a prometheus remote write sink, but I'm not sure if there's a well-known name for the pull-style sink. Exporter maybe?

Reply:
Exporter sounds great.

Reply:
Just confirming. We'll have a `prometheus_exporter` and `prometheus_remove_write` sink?

Reply:
I haven't renamed anything yet except internally. If you think that is a reasonable distinction, I'll do the rename. Is it possible to set up an alias so existing configs continue to work?

Reply:
Yes, we should set up an alias and not break backward compatibility. I would prefer those names for the distinction. cc/ @jamtur01